File: 0606-construct-string-from-binary-tree/0606-construct-string-from-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    string tree2str(TreeNode* root) {
        stack<TreeNode*> st{{root}};
        set<TreeNode*> visited;
        string ans = "";
        
        while (!st.empty()){
            TreeNode* node = st.top();
            if (visited.count(node)){
                st.pop();
                ans.push_back(')');
            }
            else{
                visited.insert(node);
                ans.push_back('(');
                ans += to_string(node -> val);
                if (!node -> left && node -> right)
                    ans += "()";
                if (node -> right)
                    st.push(node -> right);
                if (node -> left)
                    st.push(node -> left);
            }
        }
        
        return ans.substr(1, ans.size() - 2);
    }
};

File: 0606-construct-string-from-binary-tree/0606-construct-string-from-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def tree2str(self, root: Optional[TreeNode]) -> str:
        if not root:
            return ""
        stack, ans, visited = [root], "", set()
        
        while stack:
            node = stack[-1]
            if node in visited:
                stack.pop()
                ans += ')'
            else:
                visited.add(node)
                ans += '(' + str(node.val)
                if not node.left and node.right:
                    ans += "()"
                if node.right:
                    stack.append(node.right)
                if node.left:
                    stack.append(node.left)
            
        return ans[1:-1]